pkg/server/endpoints/node/handler.go

```diff
@@ -797,15 +821,94 @@ func (h *Handler) signCSRs(ctx context.Context,
 }
 
+func (h *Handler) signCSRs(ctx context.Context,
+	peerCert *x509.Certificate, csrs map[string][]byte, regEntries []*common.RegistrationEntry) (
+	svids map[string]*node.X509SVID, err error) {
+
+	callerID, err := getSpiffeIDFromCert(peerCert)
+	if err != nil {
+		return nil, err
+	}
+
+	//convert registration entries into a map for easy lookup
+	regEntriesMap := make(map[string]*common.RegistrationEntry)
+	for _, entry := range regEntries {
+		regEntriesMap[entry.EntryId] = entry
+	}
+
+	ds := h.c.Catalog.GetDataStore()
+	svids = make(map[string]*node.X509SVID)
+	//iterate the CSRs and sign them
+	for entryID, csr := range csrs {
+		spiffeID, err := getSpiffeIDFromCSR(csr, idutil.AllowAny())
+		if err != nil {
+			return nil, err
+		}
+
+		baseSpiffeIDPrefix := fmt.Sprintf("%s/spire/agent", h.c.TrustDomain.String())
+
+		sourceAddress := "unknown"
+		if peerAddress, ok := getPeerAddress(ctx); ok {
+			sourceAddress = peerAddress.String()
+		}
+
+		signLog := h.c.Log.WithFields(logrus.Fields{
+			"caller_id":      callerID,
+			"spiffe_id":      spiffeID,
+			"source_address": sourceAddress,
+		})
+
+		if spiffeID == callerID && strings.HasPrefix(callerID, baseSpiffeIDPrefix) {
+			res, err := ds.FetchAttestedNode(ctx,
+				&datastore.FetchAttestedNodeRequest{SpiffeId: spiffeID},
+			)
+			if err != nil {
+				return nil, err
+			}
+			// attested node discrepancies are not likely since the agent
+			// certificate is checked against the attested nodes during the
+			// authentication step. however, it is possible that an agent is
+			// evicted between authentication and here so these checks should
+			// remain.
+			if res.Node == nil {
+				return nil, errors.New("no record of attested node")
+			}
+			if res.Node.CertSerialNumber != peerCert.SerialNumber.String() {
+				return nil, errors.New("SVID serial number does not match")
+			}
+
+			signLog.Debug("Renewing agent SVID")
+			svid, svidCert, err := h.buildBaseSVID(ctx, csr)
+			if err != nil {
+				return nil, err
+			}
+			svids[entryID] = svid
+
+			if err := h.updateAttestationEntry(ctx, svidCert); err != nil {
+				return nil, err
+			}
+		} else {
+			signLog.Debug("Signing SVID")
+			svid, err := h.buildSVID(ctx, entryID, regEntriesMap, csr)
+			if err != nil {
+				return nil, err
+			}
+			svids[entryID] = svid
+		}
+	}
+
+	return svids, nil
+}
+
 func (h *Handler) buildSVID(ctx context.Context,
-	spiffeID string, regEntries map[string]*common.RegistrationEntry, csr []byte) (
+	ID string, regEntries map[string]*common.RegistrationEntry, csr []byte) (
 	*node.X509SVID, error) {
 
 	//TODO: Validate that other fields are not populated https://github.com/spiffe/spire/issues/161
 	//validate that is present in the registration entries, otherwise we shouldn't sign
-	entry, ok := regEntries[spiffeID]
+	entry, ok := regEntries[ID]
 	if !ok {
-		return nil, fmt.Errorf("not entitled to sign CSR for %q", spiffeID)
+		return nil, fmt.Errorf("not entitled to sign CSR for %q", ID)
```

Comment:
At the risk of making this uglier, we should probably distinguish the kind of value here. When it was only a SPIFFE ID this message was pretty obvious, but if we now emit an entry ID here, it might be confusing and hard to tell exactly what it is.

If the ID is equal to the SPIFFE ID, we can log "sign CSR for SPIFFE ID %q" and if it is not we can log "sign CSR for registration entry ID %q". Thoughts?

Reply:
Yes sure, It will make the log message clearer